# Project: Wrangle and Analyse Data

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [27]:
# Import the modules I require
import subprocess
import sys
import pandas as pd
import json
from random import randint

#### Some modules are not imported, yet
##### Why?
* they may not be available out-of-the-box, so to speak
    * this happens in older versions of python

##### Which modules, exactly?
* `tweepy`
* `requests`

##### What then?
* import said modules inside a `try`-`except` block
* a function, `install`, will fire when the `try` block fails
* flow will be as such: 
    
    ```
       try:
           import module
       except ImportError:
           install(module)
           import module
        else:
            # do something IFF `try` succeeds
        finally:
            # do something whether or not an error was thrown
            ```

In [2]:
'''
a function to install modules using `pip`

accepts one arg, type str,that is the name
of the module to install

return: None
'''
def install(mod):
    """function install"""
    subprocess.check_call([sys.executable, "-m", "pip", "install", mod])




In [3]:
#import requests

try:
    import requests
except ImportError:
    install('requests')
    import requests

if requests: 
    print(f'Successfully imported `requests` module...')

Successfully imported `requests` module...


In [4]:
#import tweepy

try:
    import tweepy
except ImportError:
    install('tweepy')
    import tweepy

if tweepy: 
    print(f'Successfully imported `tweepy` module...')

Successfully imported `tweepy` module...


In [5]:
#Constants
with open('./keys/twtr_api_keys.json', 'r') as f:
    data = json.load(f)
CONSUMER_KEY = data['api_key']
CONSUMER_SECRET = data['api_key_secret']
ACCESS_TOKEN = data['access_key']
ACCESS_SECRET = data['access_token']

In [6]:
#download twtr archive data then upload it to workspace
twtr_archive = pd.read_csv('twitter_archive_enhanced.csv', sep=',')

In [7]:
#view the file
twtr_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [ ]:
#a way to download twtr archive data, write and save it to file
# w/o physically downloading then uploading it

url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/59a4e958_twitter-archive-enhanced/twitter-archive-enhanced.csv'
r = requests.get(url)
with open('twitter_archive_enhanced.csv', 'wb') as f:
    f.write(r.content)

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [8]:
#use `requests` to download tweet image prediction data & save to file
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
im = requests.get(url)

with open('image_predictions.tsv', 'wb') as f:
    f.write(im.content)

In [9]:
#view the file
img_predictions = pd.read_csv('image_predictions.tsv', sep='\t')
img_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [10]:
#set up the api object
auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
auth.set_access_token(ACCESS_TOKEN, ACCESS_SECRET)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)


In [12]:
#query additional data via twtr api
with open('tweet_json.txt', 'a', encoding='utf8') as f:
    for i in twtr_archive['tweet_id']:
        try:
            tweet = api.get_status(i, tweet_mode='extended')
            json.dump(tweet._json, f)
            f.write('\n')
        except:
            continue

Rate limit reached. Sleeping for: 736
Rate limit reached. Sleeping for: 736


In [9]:
#append each tweet into a list
with open('tweet_json.txt', 'r') as f:
    tweets = f.readlines()
    try:
        tweet_list = [json.loads(i) for i in f]
    except:
#         continue
#         raise
        pass



In [10]:
# alternative way to append each tweet into a list
tweet_list = []

with open('tweet_json.txt', 'r') as f:
    for i in f:
        try:
            tweet = json.loads(i)
            tweet_list.append(tweet)
        except:
            continue
        

In [11]:
#check length of `tweet_list`
len(tweet_list) #6378 obs

6378

In [12]:
#create tweets to df
df_tweets = pd.DataFrame()

#add variables and values to df
df_tweets['tweet_id'] = list(map(lambda tweet: tweet['id'], tweet_list))
df_tweets['retweet_count'] = list(map(lambda tweet: tweet['retweet_count'], tweet_list))
df_tweets['favorite_count'] = list(map(lambda tweet: tweet['favorite_count'], tweet_list))

In [13]:
#save `df_tweets` as `.csv` file for ease of reference
df_tweets.to_csv('twitter_api_data.csv', index=False)

In [14]:
#check if df_tweets exists and that it is not empty
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6378 entries, 0 to 6377
Data columns (total 3 columns):
tweet_id          6378 non-null int64
retweet_count     6378 non-null int64
favorite_count    6378 non-null int64
dtypes: int64(3)
memory usage: 149.6 KB


## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



### Method

* assess data sets one by one
* perform the following on each data set:
    * under  heading `Structure`:
        * visual assessment
        * summary of programmatic assessment
        * list of properties and/or attributes
    * list of quality and tidiness issues
* summary list of quality and tidiness issues in all data sets

#### Dataset 1: Dataframe `twtr_archive` that contains archive data

In [15]:
twtr_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [16]:
twtr_archive.shape

(2356, 17)

In [17]:
twtr_archive.duplicated().value_counts()

False    2356
dtype: int64

In [18]:
twtr_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [28]:
twtr_archive.tweet_id.nunique()

2356

In [29]:
twtr_archive.sample(randint(5, 15))

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
88    875097192612077568                    NaN                  NaN   
1221  714485234495041536                    NaN                  NaN   
1854  675534494439489536                    NaN                  NaN   
1812  676811746707918848                    NaN                  NaN   
916   756998049151549440                    NaN                  NaN   
1151  725842289046749185                    NaN                  NaN   
758   778408200802557953                    NaN                  NaN   
1565  688064179421470721                    NaN                  NaN   
2129  670290420111441920                    NaN                  NaN   
1431  697516214579523584                    NaN                  NaN   

                      timestamp  \
88    2017-06-14 21:06:43 +0000   
1221  2016-03-28 16:12:09 +0000   
1854  2015-12-12 04:35:48 +0000   
1812  2015-12-15 17:11:09 +0000   
916   2016-07-23 23:42:53 +0000   
1151  2016-04-29 00:21:01 +0000   
758   2016-09-21 01:39:11 +0000   
1565  2016-01-15 18:24:18 +0000   
2129  2015-11-27 17:17:44 +0000   
1431  2016-02-10 20:23:19 +0000   

                                                 source  \
88    <a href="http://twitter.com/download/iphone" r...   
1221  <a href="http://vine.co" rel="nofollow">Vine -...   
1854  <a href="http://twitter.com/download/iphone" r...   
1812  <a href="http://twitter.com/download/iphone" r...   
916   <a href="http://twitter.com/download/iphone" r...   
1151  <a href="http://twitter.com/download/iphone" r...   
758   <a href="http://twitter.com/download/iphone" r...   
1565  <a href="http://twitter.com/download/iphone" r...   
2129  <a href="http://twitter.com/download/iphone" r...   
1431  <a href="http://vine.co" rel="nofollow">Vine -...   

                                                   text  retweeted_status_id  \
88    You'll get your package when that precious man...                  NaN   
1221  This is Blitz. He screams. 10/10 (vid by @yeaa...                  NaN   
1854  Seriously guys?! Only send in dogs. I only rat...                  NaN   
1812  Say hello to Penny &amp; Gizmo. They are pract...                  NaN   
916   This is Oliver. He's an English Creamschnitzel...                  NaN   
1151  This is Colby. He's currently regretting all t...                  NaN   
758   RIP Loki. Thank you for the good times. You wi...                  NaN   
1565  This is Kilo. He's a Pouncing Brioche. Really ...                  NaN   
2129  This is Sandra. She's going skydiving. Nice ad...                  NaN   
1431  Say hello to Pherb. He does parkour. 9/10 http...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
88                         NaN                        NaN   
1221                       NaN                        NaN   
1854                       NaN                        NaN   
1812                       NaN                        NaN   
916                        NaN                        NaN   
1151                       NaN                        NaN   
758                        NaN                        NaN   
1565                       NaN                        NaN   
2129                       NaN                        NaN   
1431                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
88    https://twitter.com/drboondoc/status/874413398...                13   
1221                      https://vine.co/v/iDrOvVqq0A6                10   
1854  https://twitter.com/dog_rates/status/675534494...                11   
1812  https://twitter.com/dog_rates/status/676811746...                 9   
916   https://twitter.com/dog_rates/status/756998049...                11   
1151  https://twitter.com/dog_rates/status/725842289...                12   
758   https://twitter.com/dog_rates/status/778408200...                14   
1565  https://twit

In [ ]:
twtr_archive

#### Structure

##### Overall

* 2356 observations
    
* 17 variables:
    * 4 of type `float`
    * 3 of type `int`
    * 10 of type `str`
* time stamps are in `str` format:
    * `timestamp`
    * `retweeted_status_timestamp`
* 5 unique IDs:
    * `tweet_id`
    * `in_reply_to_status_id`
    * `in_reply_to_user_id`
    * `retweeted_status_id`
    * `retweeted_status_user_id`
* the age range of a dog is presented qualitatively using the variables:
    * `puppo` -> very young dog
    * `puppy` -> young-to-adolescent dog
    * `doggo` -> adult dog
* the `floofer` variable might represent the fur type and/or fur texture of a dog
    

##### Missing and null values
* 6 out of 17 variables have null values
* Case(s) in point:
    * 2278 values missing in the `in_reply_to_status_id` and `in_reply_to_user_id` variables
    * 2175 values missing in the `retweeted_status_id`, `retweeted_status_user_id`  and `retweeted_status_timestamp` variables 

##### Duplicated observations
* `twtr_archive` has no duplicated observations

##### Multiple values for a variable
* observations in `twtr_archive` have a single value per variable
     

##### Quality and tidiness

###### Quality
* missing values for observations
* some values under the `name` variable do not fit in
    * examples: 
        * value `a` in observation `2354`
        * value `such` in observation `22`
* the case of values under the `name` variable is not consistent
    * some values are all lower case
    * the first letter of some is upper case
* some values under the `rating_denominator` variable are not `10`
    * the value of observation `2335`, for example, is `2`


###### Tidiness
* 
* 

#### Dataset 2: Dataframe `img_predictions` that contains image data

In [19]:
img_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [20]:
img_predictions.shape

(2075, 12)

In [21]:
img_predictions.duplicated().value_counts()

False    2075
dtype: int64

In [22]:
img_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [30]:
img_predictions.tweet_id.nunique()

2075

In [31]:
img_predictions.sample(randint(5, 15))

tweet_id                                          jpg_url  \
874   698178924120031232  https://pbs.twimg.com/media/CbBuBhbWwAEGH29.jpg   
325   671882082306625538  https://pbs.twimg.com/media/CVMBL_LWUAAsvrL.jpg   
198   669625907762618368  https://pbs.twimg.com/media/CUr9NjgU8AEpf5w.jpg   
165   668979806671884288  https://pbs.twimg.com/media/CUixld6WoAArDrJ.jpg   
60    667138269671505920  https://pbs.twimg.com/media/CUImtzEVAAAZNJo.jpg   
1354  760252756032651264  https://pbs.twimg.com/media/Coz12OLWgAADdys.jpg   
255   670755717859713024  https://pbs.twimg.com/media/CU8AwZ_UsAA-Lbu.jpg   
956   705428427625635840  https://pbs.twimg.com/media/CcovaMUXIAApFDl.jpg   
1660  811627233043480576  https://pbs.twimg.com/media/C0N6opSXAAAkCtN.jpg   
1388  766313316352462849  https://pbs.twimg.com/media/CqJ95SRWgAATPK_.jpg   
565   678278586130948096  https://pbs.twimg.com/media/CWm6xySUEAAqfFU.jpg   
511   676191832485810177  https://pbs.twimg.com/media/CWJQ4UmWoAIJ29t.jpg   

      img_num                           p1   p1_conf  p1_dog  \
874         1     Chesapeake_Bay_retriever  0.351868    True   
325         1                     ski_mask  0.968325   False   
198         1                    seat_belt  0.874502   False   
165         1             golden_retriever  0.608537    True   
60          1  West_Highland_white_terrier  0.747713    True   
1354        1              radio_telescope  0.155279   False   
255         1                     keeshond  0.994065    True   
956         1                    Chihuahua  0.774792    True   
1660        1                       beagle  0.396280    True   
1388        1                   toy_poodle  0.966896    True   
565         1                  Maltese_dog  0.897841    True   
511         2                    Chihuahua  0.376741    True   

                      p2   p2_conf  p2_dog                       p3   p3_conf  \
874             malinois  0.207753    True       Labrador_retriever  0.154606   
325                 mask  0.021863   False                    abaya  0.005479   
198     golden_retriever  0.055408    True       Labrador_retriever  0.026854   
165         Irish_setter  0.097078    True                  redbone  0.076022   
60               Samoyed  0.243629    True               toy_poodle  0.001804   
1354                 dam  0.154515   False                    crane  0.098040   
255   Norwegian_elkhound  0.001827    True                    cairn  0.001821   
956                quilt  0.073079   False                 Pembroke  0.022365   
1660            Pembroke  0.049562    True  wire-haired_fox_terrier  0.046349   
1388    miniature_poodle  0.016424    True           cocker_spaniel  0.010227   
565                Lhasa  0.035717    True          Tibetan_terrier  0.017107   
511    Italian_greyhound  0.173114    True                   muzzle  0.071485   

      p3_dog  
874     True  
325    False  
198     True  
165     True  
60      True  
1354   False  
255     True  
956     True  
1660    True  
1388    True  
565     True  
511    False

In [ ]:
img_predictions

#### Structure

##### Overall

* 2075 observations   
* 12 variables:
    * 3 of type `float`
    * 2 of type `int`
    * 4 of type `str`
    * 3 of type `bool`
* 1 unique ID: `tweet_id`
* the predicted breed of a dog in a photo is presented using the variable `pN`, where `N` is 1, 2 or 3
    * example: `p1` -> prediction 1
* the likelihood that the breed of a dog is the value of variable `pN` is presented using the variable `pN_conf`
* whether or not the image at `jpg_url` is that of a dog is presented under the variable `pN_dog`    

##### Missing and null values
* `img_predictions` has no missing and/or null values 

##### Duplicated observations
* `img_predictions` has no duplicated observations

##### Multiple values for a variable
* observations in `img_predictions` have a single value per variable
     

##### Quality and tidiness

###### Quality
* values under variables `pN` (where `N` is 1, 2 or 3) do not have a  consistent case
    * examples:
        * value `redbone` in observation `1` under variable `p1`
        * value `Ibizan_hound` in observation `12` under variable `p3`
* some values under variables `pN` do not fit in; that is to say, they are not breeds of dogs
    * examples:
        * value `three-toed_sloth` in observation `21` under variable `p1`
        * value `skunk` in observation `25` under variable `p2`
* some observations do not fit in
    * examples:
        * observation `17` appears to be about birds
        * observation `18` appears to be about library or office equipment
        * observation `25` appears to be about small mammals
    * however, the values of said observations under variables `pN_dog` are `False`. The classifier identified that the respective images in question were not of dogs

* some variable names are not descriptive
    * notice how "`pN` (where `N` is 1, 2 or 3)" is used to refer to `p1`, `p2` and `p3`
       * saves a few keystrokes but is difficult for anyone that assess the data set. `pred_N` or `prediction_N` are more descriptive 
* 
* 
* 
* 

###### Tidiness
* 
* 

#### Dataset 3: Dataframe `df_tweets` that contains twitter API data

In [23]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6378 entries, 0 to 6377
Data columns (total 3 columns):
tweet_id          6378 non-null int64
retweet_count     6378 non-null int64
favorite_count    6378 non-null int64
dtypes: int64(3)
memory usage: 149.6 KB


In [24]:
df_tweets.shape

(6378, 3)

In [25]:
df_tweets.duplicated().value_counts()

False    3991
True     2387
dtype: int64

In [26]:
df_tweets.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           6981           33730
1  892177421306343426           5281           29261
2  891815181378084864           3468           21992
3  891689557279858688           7202           36833
4  891327558926688256           7724           35219

In [32]:
df_tweets.tweet_id.nunique()

2327

In [ ]:
df_tweets

#### Structure

##### Overall

* 6378 observations
    
* 3 variables; all of type `float`
* 1 unique ID: `tweet_id`
* `retweet_count` shows the number of re-tweets
* `favorite_count` shows the number of times the tweet under `tweet_id` has been marked as favourite
    

##### Missing and null values
* `df_tweets` has no missing and/or null values 

##### Duplicated observations
* `df_tweets` has 2387 duplicated observations
* variable `tweet_id` has only 2327 unique values 

##### Multiple values for a variable
* observations in `df_tweets` have a single value per variable
     

##### Quality and tidiness

###### Quality
* almost half of the obsevations are duplicates
* 
* 
* 
* 
* 
* 
* 

###### Tidiness
* 
* 

#### Summary of issues

#### Quality

* missing values for observations
    * example from `twtr_archive`: 2278 values missing in the `in_reply_to_status_id` variable

* some values do not fit in
    * example from `twtr_archive`: value `such` in observation `22`

* the case of values under the some variable is not consistent
    * example from `img_predictions`: value `Ibizan_hound` in observation `12` under variable `p3`

* some values under the `rating_denominator` variable in `twtr_archive` are not `10`
    * the value of observation `2335`, for example, is `2`

* some variable names in `img_predictions` are not descriptive
    * "`pN` (where `N` is 1, 2 or 3)" is used to refer to `p1`, `p2` and `p3`

* `img_predictions`: some values under variables `pN` do not fit in; that is to say, they are not breeds of dogs
    * example: value `three-toed_sloth` in observation `21` under variable `p1`

* some observations do not fit in
    * example from `img_predictions`: observation `17` appears to be about birds
    * the values of said observations under variables `pN_dog`, however,  are `False`. The classifier identified that the respective images in question were not of dogs

* `df_tweets`: almost half of the observations are duplicates

##### Tidiness

## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [ ]:
# Make copies of original pieces of data


### Issue #1:

#### Define:

#### Code

#### Test

### Issue #2:

#### Define

#### Code

#### Test

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

### Insights:
1.

2.

3.

### Visualization